keps/sig-storage/20191012-storagepool-advplacement.md

```diff
@@ -0,0 +1,363 @@
+```
+type StoragePool struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+ 
+        // Spec defines the desired state of a storage pool
+        // +optional
+        Spec StorgePoolSpec
+ 
+        // Status represents the current information about a storage pool
+        // +optional
+        Status StoragePoolStatus
+}
+ 
+// StoragePoolSpec describes the attributes of a storage pool
+Type StoragePoolSpec struct {
+    	// Name of the driver
+    	Driver string
+ 
+    	// Opaque parameters describing attributes of the storage pool
+    	+optional
+    	Parameters map[string]string
+}
+ 
+type StoragePoolStatus struct {
+    	// Nodes the storage pool has access to
+    	+optional
+  	AccessibleNodes []string
+  	// Capacity of the storage pool
+  	+optional
+  	Capacity *PoolCapacity
+  	// Last errors happened on the pool
+  	+optional
+  	Errors []StoragePoolError
+}
+ 
+type PoolCapacity struct {
+    	// Total capacity of the storage pool
+    	// +optional
+    	Total *resource.Quantity
+}
+ 
+// Describes an error encountered on the pool
+type StoragePoolError struct {
+    	// Time is the timestamp when the error was encountered.
+    	// +optional
+    	Time *metav1.Time
+ 
+    	// Message details the encountered error
+    	// +optional
+    	Message *string
+}
+```
+
+Note that `StoragePool` are non-namespaced. 
+
+#### CSI changes
+
+* Add a ControllerListStoragePools RPC that lists storage pools for CSI drivers that support LIST_STORAGE_POOLS controller capability.
+
+#### New external controller for StoragePool
+
+CSI will be extended with a new ControllerListStoragePools() API, which returns all storage pools available via this CSI driver. A new external K8s controller uses this API upon CSI startup to learn about all storage pools and publish them into the K8s API server using the new StoragePool CRD. 
+
+#### Usage of StoragePool objects by application
+Storage applications (e.g. MongoDB, Kafka, Minio, etc.), with their own operator (i.e. controllers), consume the data in the StoragePool. It is how they understand topology. For example, let’s say we are in User Story 2, i.e. every Node has a bunch of local drives each published as a StoragePool. The operator will understand that they are “local” drives by seeing that only one K8s Node has access to each of them, and it will see how many there are and of which size. Based on the mirroring/erasure coding scheme of the specific storage service, it can now translate to how many PVCs, which size and on which of these StoragePools to create. Given that the StoragePool may model something as fragile as a single physical drive, it is a real possibility for a StoragePool to fail or be currently inaccessible. The StoragePool hence also communicates that fact, and the Storage Service operator can understand when and why volumes are failing, and how to remediate (e.g. by putting an additional PVC on another pool and moving some data around). 
```

Comment:
Instead of one storagepool per local drive, is there possibility of grouping set of local drives which then makes sense as pool?
If app operator need to really use storagePool which is one per local drive, instead, it can actually create a PV out of this local drive and use the same in app deployment/STS.